p2p/protocol/identify/id.go

```diff
@@ -146,40 +154,96 @@ func NewIDService(h host.Host, opts ...Option) *IDService {
 
 	// register protocols that do not depend on peer records.
 	h.SetStreamHandler(IDDelta, s.deltaHandler)
-	h.SetStreamHandler(LegacyID, s.requestHandler)
-	h.SetStreamHandler(LegacyIDPush, s.pushHandler)
-
-	// register protocols that depend on peer records.
-	h.SetStreamHandler(ID, s.requestHandler)
+	h.SetStreamHandler(ID, s.sendIdentifyResp)
 	h.SetStreamHandler(IDPush, s.pushHandler)
 
 	h.Network().Notify((*netNotifiee)(s))
 	return s
 }
 
-func (ids *IDService) handleEvents() {
+func (ids *IDService) loop() {
 	defer ids.refCount.Done()
 
+	phs := make(map[peer.ID]*peerHandler)
 	sub, err := ids.Host.EventBus().Subscribe([]interface{}{&event.EvtLocalProtocolsUpdated{},
 		&event.EvtLocalAddressesUpdated{}}, eventbus.BufSize(256))
 	if err != nil {
 		log.Errorf("failed to subscribe to events on the bus, err=%s", err)
 		return
 	}
 
-	defer sub.Close()
+	defer func() {
+		sub.Close()
+		for pid := range phs {
```

Comment:
nit:

```go
for _, ph := range phs {
  ph.close()
}
```

Reply:
@Stebalien I really like the current one better because I've run into golang problems when iterating over a map where the value type is a pointer(though that isn't the case here, I now prefer that style).

Reply:
Eh, I don't mind much either way. I just find it cleaner to iterate by value (also likely slightly faster but that _really_ doesn't matter here).

What was your bad experience?